src/carts/src/main/java/mushop/carts/Cart.java

```diff
@@ -50,9 +51,24 @@ public boolean removeItem(String itemId) {
 
     public void merge(Cart cart) {
         this.customerId = cart.getCustomerId();
-        for (Item item : cart.items) {
-            items.add(item);
-        }
+        List<Item> existingItems = new ArrayList<Item>(this.items);
+        cart.items.forEach((item) -> {
```

Comment:
Thanks for fixing the bugs and sorry to snoop on your change. 

I assume the fix here is to avoid duplicate items.  i.e (1) if the item already exists, then update its quantity and (2) if the item id does not exist, add it.  

Currently it is making a copy of the existing items before iterating the incoming items.  For each item, you are calling contains() which for ArrayList is going to iterate the internal array to test if its there.  Then, you will iterate the existingItems again when you call forEach.   Also, so that the contains has has the correct semantics, you have added hashCode and equals to Item.  (btw, ArrayList doesn't use hashCode :
https://docs.oracle.com/javase/8/docs/api/java/util/List.html#contains-java.lang.Object- )

It just occurred to me that you could avoid hashCode, equals, the copy, and these extra iterations if you do something like this:

```
public void merge(Cart cart) {
   for (Item item : cart.items) {
      mergeItem(item);
   }
}

private void mergeItem(Item item) {
   for (Item existing : items) {
      if (existing.getId() == item.getId()) {
         existing.setQuantity(.....);
         return;
      }
   }
   existing.add(item);
}
```

Also, should it be existing.getQuantity() + item.getQuanity()?

Reply:
Yes thats concise and much more efficient ! Thanks !
I'm updating the PR to use this

Reply:
BTW - I dont think you are in this github org? I could not assign you as a reviewer.

Reply:
@jjspiegel - Actually it will result in concurrent modification. The same thread that is iterating over the List in `merge` would be trying to add to the List in `mergeItem`, and that would throw it off with a runtime exception.

May be a better way is to not use `.forEach` and do a very traditional indexed iteration, since these are arraylists and index access is more efficient.